src/controllers/plugins/api/HTTPRequest.cpp

```diff
@@ -6,402 +6,183 @@
-                             QByteArray::fromStdString(value));
-    return 0;
-}
-
-int HTTPRequest::create(lua_State *L)
-{
-    lua::StackGuard guard(L, -1);
-    if (lua_gettop(L) != 2)
-    {
-        return luaL_error(
-            L, "HTTPRequest.create needs exactly 2 arguments (method "
-               "and url)");
-    }
-    QString url;
-    if (!lua::pop(L, &url))
-    {
-        return luaL_error(L,
-                          "cannot get url (2nd argument of HTTPRequest.create, "
-                          "expected a string)");
-    }
     auto parsedurl = QUrl(url);
     if (!parsedurl.isValid())
     {
-        return luaL_error(
-            L, "cannot parse url (2nd argument of HTTPRequest.create, "
-               "got invalid url in argument)");
-    }
-    NetworkRequestType method{};
-    if (!lua::pop(L, &method))
-    {
-        return luaL_error(
-            L, "cannot get method (1st argument of HTTPRequest.create, "
-               "expected a string)");
+        throw std::runtime_error(
+            "cannot parse url (2nd argument of HTTPRequest.create, "
+            "got invalid url in argument)");
     }
     auto *pl = getApp()->getPlugins()->getPluginByStatePtr(L);
     if (!pl->hasHTTPPermissionFor(parsedurl))
     {
-        return luaL_error(
-            L, "Plugin does not have permission to send HTTP requests "
-               "to this URL");
+        throw std::runtime_error(
+            "Plugin does not have permission to send HTTP requests "
+            "to this URL");
     }
     NetworkRequest r(parsedurl, method);
-    lua::push(
-        L, std::make_shared<HTTPRequest>(ConstructorAccessTag{}, std::move(r)));
-    return 1;
-}
-
-int HTTPRequest::execute_wrap(lua_State *L)
-{
-    auto ptr = HTTPRequest::getOrError(L, 1);
-    return ptr->execute(L);
+    return std::make_shared<HTTPRequest>(ConstructorAccessTag{}, std::move(r));
 }
 
-int HTTPRequest::execute(lua_State *L)
+void HTTPRequest::execute(sol::this_state L)
 {
-    auto shared = this->shared_from_this();
+    if (this->done)
+    {
+        throw std::runtime_error(
+            "Cannot execute this c2.HTTPRequest, it was executed already!");
+    }
     this->done = true;
-    std::move(this->req_)
-        .onSuccess([shared, L](const NetworkResult &res) {
-            lua::StackGuard guard(L);
-            auto *thread = lua_newthread(L);
 
-            auto priv = shared->pushPrivate(thread);
-            lua_getfield(thread, priv, "success");
-            auto cb = lua_gettop(thread);
-            if (lua_isfunction(thread, cb))
+    // this keeps the object alive even if Lua were to forget about it,
+    auto hack = this->weak_from_this();
+    auto *pl = getApp()->getPlugins()->getPluginByStatePtr(L);
+    pl->httpRequests.push_back(this->shared_from_this());
+
+    std::move(this->req_)
+        .onSuccess([this, L, hack](const NetworkResult &res) {
+            if (hack.expired())
```

Comment:
I'd prefer if we use lock on hack here, rather than checking expired

Reply:
I will change it however I don't believe this matters as all of this is expected to run on a single thread, the object won't be deleted mid-callback.